### ISSUE TITLE
EditContext: Who's going to handle caret navigation?

### ISSUE THREAD

**Issue by BoCupp-Microsoft** [w3c/editing#266]
Moving [an EditContext issue](https://github.com/MicrosoftEdge/MSEdgeExplainers/issues/118) from @Reinmar into this repo.

## Problem
On the Editor TF meeting in Paris, back in 2015, we talked a lot about the challenges in handling caret navigation. There are dozens of interactions to handle:

* up/right/down/left arrow keys
* jumping over one word
* jumping to the EOL (home/end)
* shift + the above
* ctrl + home/end
* page up/down
* mouse
* touch
* pen?
* and whatever else will appear in the future or exists already

To add to that, we have mixed RTL and LTR content which is extremely tricky to handle, we have OS specific behaviours, etc.

In other words, it's a very small chance that a JS app will get all those things right. And we agreed on that in Paris.

Reading the EditContext explainer I found an example like:

```javascript
editContainer.addEventListener("keydown", e => {
    // Handle control keys that don't result in characters being inserted
    switch (e.key) {
        case "Home":
            model.updateSelection(...);
            view.queueUpdate();
            break;
        case "Backspace":
            model.deleteCharacters(Direction.BACK);
            view.queueUpdate();
            break;
        ...
    }
});
```
And it seems rather unrealistic. For three reasons:

* "intention" detection will be tricky to get right (OS-specific, key combinations, etc.)
* calculating the results for all the interactions is extremely tricky (page up/down, arrow keys themselves)
* mouse + touch + pen + whatever else come to your mind – you just don't want to do that in JS

## Questions
* Do I get the intention of this explainer right? Was it meant to push all the responsibility to the JS side? I'm asking because at the end it briefly mentions some alternatives, but it's unclear whether these are alternatives to the entire EditContext API or e.g. alternative ways to handle user interactions.
* Followup question – couldn't we have the browser firing events (just like beforeInput and perhaps berforeSelectionChange) for those user interactions mentioned in this ticket? They would not only provide a better intention detection than listening to keydown or mousemove, but can also indicate to the JS app what the browser would normally do itself (by targetRanges e.g.).

## PS
This is my biggest doubt regarding EditContext. The other one is spell checking (will open a separate ticket in a moment). Other than that, I think it's an interesting direction. But a key point for me here is making this API "adoptable" for less complex apps than Google Docs and Office Online. So, apps which do not calculate the entire layout themselves.

**Comment by BoCupp-Microsoft** [comment]
> * "intention" detection will be tricky to get right (OS-specific, key combinations, etc.)

Our latest thinking is that the EditContext can now be associated with one or more DOM elements.  The association allows certain editable behaviors to be enabled on the active, EditContext-associated element.  One of those behaviors is the dispatching of a `beforeinput` event, which communicates the user's intent.  

I think that will address this concern, but if there are user intentions relevant to editing that are not covered by `beforeinput` events please call those out specifically and we can discuss how to detect them. 

> * calculating the results for all the interactions is extremely tricky (page up/down, arrow keys themselves)

Content-based navigation can be handled by the EditContext.  In [beforeinput terms](https://www.w3.org/TR/input-events-1/#interface-InputEvent-Attributes) this includes the selection location after these editing operations:
* insertText
* insertReplacementText
* insertLineBreak
* insertParagraph
* insertTranspose
* insertCompositionText
* deleteWordBackward
* deleteWordForward
* deleteHardLineBackward
* deleteHardLineForward
* deleteContent
* deleteContentBackward
* deleteContentForward

And also includes the selection location after the user initiates these explicit caret/selection movements (I've just extracted these names from the Chromium list of [editing command types here](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/editing/commands/editing_command_type.h;bpv=1;bpt=1;l=14) to try to be complete):
* moveBackward
* moveForward
* moveLeft
* moveParagraphBackward
* moveParagraphForward
* moveRight
* moveToBeginningOfDocument
* moveToBeginningOfParagraph
* moveToBeginningOfSentence
* moveToEndOfDocument
* moveToEndOfParagraph
* moveToEndOfSentence
* moveWordBackward
* moveWordForward
* moveWordLeft
* moveWordRight

The location of selection would be reported in the `textupdate` event dispatched to the `EditContext` and the author would then translate the text-based offsets describing the selection into their own document model and view.

Layout-based navigation **cannot** be handled by the EditContext (at least not as currently defined).  These would be the responsibilty of the author:
* moveDown
* movePageDown
* movePageUp
* moveToBeginningOfLine
* moveToEndOfLine
* moveToLeftEndOfLine
* moveToRightEndOfLine
* moveUp

Some editing operations from the [InputEvents specification](https://www.w3.org/TR/input-events-1/#interface-InputEvent-Attributes) also require layout information which **cannot** be currently handled by the EditContext:
* deleteSoftLineBackward
* deleteSoftLineForward
* deleteEntireSoftLine

The biggest impediment to implementing operations that require layout information is the lack of it.  We could either introduce APIs to query line layout information to help with the implementation of these editing operations, or it may be sufficient to just provide an API like [Selection.modify](https://developer.mozilla.org/en-US/docs/Web/API/Selection/modify) - given a caret position you can find another in a given direction and for a given unit.  

Note that it is also getting easier to implement line layout with proposals like [Canvas Formatted Text](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/Canvas/FormattedText.md).

> * mouse + touch + pen + whatever else come to your mind – you just don't want to do that in JS

If you implement your own selection, like Monaco, Word, GDocs, then these events are the responsibility of the author.  I read this point as wanting to use native selection and the caret with editable regions that are enabled via EditContext.  I don't have an answer for how that will work yet, but will try to define something.

**Comment by johanneswilm** [comment]
I agree with @reinmar that EditContext is not doing what is needed in terms of caret movement. But it was my understanding that one could couple EditContext with `contenteditbale=caret-only` or similar to let the browser take care of selection movement. And until such a mode exists, it should work with `contenteditable=true` where one rolls back any change that the browser may then make to the dom element (before reapplying what it really needs to do), so that one effectively created a `caret-only` mode.

Background: Creating your own selection entirely in JS for a richtext editor is difficult, yet necessary in some cases. JS editors seem to mostly work with a combination where it generally uses the browser native selection, but in some cases (such as the user selecting a column of a table), it overrides with it's own selection-mechanism.
